concepts/functions/introduction.md

```diff
@@ -1,46 +1,113 @@
+  return "Hello " + name
+}
+
+func HelloAndGoodbye(name string) (string. string) {
+  return "Hello " + name, "Goodbye " + name
+}
 ```
 
-The value of a variable can be assigned using the `:=` and updated using the `=`. Once defined, a variable's type can never change.
+## Invoking Functions
+
+Invoking a function is done by specifying the function name and passing arguments for each of the function's parameters in parenthesis.
 
 ```go
-count := 1 // Assign initial value
-count = 2  // Update to new value
+import "fmt"
 
-// Compiler error when assigning different type
-// count = false
+// No parameters, no return value
+func PrintHello() {
+    fmt.Println("Hello")
+}
+// Called like this:
+PrintHello()
+
+// One parameter, one return value
+func Hello(name string) string {
+  return "Hello " + name
+}
+// Called like this:
+greeting := Hello("Dave")
+
+// Multiple parameters, multiple return values
+func SumAndMultiply(a, b int) (int, int) {
+    return a+b, a*b
+}
+// Called like this:
+aplusb, atimesb := SumAndMultiply(a, b)
 ```
 
-A function can have zero or more parameters. All parameters must be explicitly typed, there is no type inference for parameters. A function can also have multiple return values which must also be explicitly typed. Values are returned from functions using the `return` keyword. To allow a function to be called by code in other packages, the name of the function must start with a capital letter.
+## Named Return Values and Naked Return
 
-```go
-package greeting
+As well as parameters, return values can optionally be named.
+If named return values are used, a `return` statement without arguments will return those values, this is known as a 'naked' return.
 
-// Hello is a public function
-func Hello(name string) string {
-    return hello(name)
+```go
+func SumAndMultiplyThenMinus(a, b, c int) (sum, mult int) {
+    sum, mult = a+b, a*b
+    sum -= c
+    mult -= c
+    return
 }
+```
+
+## Pointers
 
-// hello is a private function
-func hello(name string) string {
-    return "Hello " + name
+Functions in go pass their arguments by value.
+This means that the arguments are copied, and any changes to those arguments within the function will not be seen outside the function.
+
+```go
+val := 2
+func MultiplyByTwo(v int) int {
+    v = v * 2
+    return v
 }
+newval := MultiplyByTwo(val)
+// newval is 4, val is 2
 ```
 
-Invoking a function is done by specifying the function name and passing arguments for each of the function's parameters.
+To affect the values a function is called with, it is possible to pass pointer arguments, using the `*` syntax in the function, and the `&` syntax to generate a pointer.
+
+```go
+func swap(x, y *int) {
+    var temp int
+    temp = *x
+    *x = *y
+    *y = temp
+}
 
-Go supports two types of comments. Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
+var a int = 100
+var b int = 200
+swap(&a, &b)
+// a = 200, b = 100
+```
```

Comment:
This section can be removed as this exercise should not cover pointers (nor is it needed by the student to solve the exercise).

Reply:
Damn, I knew I'd end up leaving something extraneous in the introduction, thanks for spotting!